### PR TITLE
plan: refactor projection elimination

### LINKED ISSUE #3681: refactor projection elimination
1. What did you do?

2. What did you expect to see?

As mentioned in this pull request [https://github.com/pingcap/tidb/pull/3625](https://github.com/pingcap/tidb/pull/3625), the `FromID` of every ancestor's  column referenced to the removed parent should be replaced with the removed parent's child's `FromID` respectively

3. What did you see instead?

4. What version of TiDB are you using (`tidb-server -V`)?

### PR DESCRIPTION
**Pull request by zz-jason**
fix #3681

### CODE REVIEW

**Comment by zz-jason** [comment]
@hanfei1991 @XuHuaiyu PTAL

**Comment by hanfei1991** [comment]
@zz-jason Can we move this optimization to logical optimizations ?

**Comment by zz-jason** [comment]
@hanfei1991 As the last stage of logical optimization ?

**Comment by hanfei1991** [comment]
I think it's ok to add after column prunning.

File: expression/column.go [hunk 1]
```diff
@@ -146,6 +146,10 @@ func (col *Column) String() string {
 	return result
 }
 
+func (col *Column) Identify() string {
```
Comment (hanfei1991) [inline expression/column.go]: We don't need a `Identify`. You can use `HashCode()` directly.

File: plan/eliminate_projection.go [hunk 2]
```diff
@@ -14,50 +14,215 @@
...
 	}
+	childCols := child.Schema().Columns
 	for i, expr := range p.Exprs {
-		col, ok := expr.(*expression.Column)
-		if !ok {
-			return false
-		}
-		if col.FromID != child.Schema().Columns[i].FromID || col.Position != child.Schema().Columns[i].Position {
+		projCol, ok := expr.(*expression.Column)
+		if !ok || projCol.Identify() != childCols[i].Identify() {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Column has a `Equal` interface

File: plan/eliminate_projection.go [hunk 3]
```diff
@@ -14,50 +14,215 @@
...
 }
+
+func resolveColumnAndReplace(origin *expression.Column, replace map[string]*expression.Column) {
+	dst := replace[(*origin).Identify()]
+	if dst != nil {
+		*origin = *dst
+	}
+}
+
+func resolveExprAndReplace(origin expression.Expression, replace map[string]*expression.Column) {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: There is a function named `expression.ColumnSubstitute` can do this.

Comment (zz-jason) [inline plan/eliminate_projection.go]: If use `expression.ColumnSubstitute(expr Expression, schema *Schema, newExprs []Expression)`:
1. We have to convert `replace`, whose type is `map[string]*expression.Column`, to a `(schema *Schema`, `newExprs [] Expression)` pair every time before function `expression.ColumnSubstitute` called. It's inconvenient
 2. `expression.ColumnSubstitute` uses a `for` loop to find whether a column exists in `schema *Schema`. It's inefficient

I prefer not to use the function `expression.ColumnSubstitute`

Comment (hanfei1991) [inline plan/eliminate_projection.go]: ok

File: expression/column.go (hunk 1 above)
Comment (zz-jason) [inline expression/column.go]: done

File: plan/eliminate_projection.go (hunk 2 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: done

File: plan/eliminate_projection.go (hunk 3 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: @hanfei1991 updated, PTAL

File: plan/eliminate_projection.go [hunk 4]
```diff
@@ -14,50 +14,215 @@
...
-	}
-	children := make([]Plan, 0, len(p.Children()))
-	for _, child := range p.Children() {
-		children = append(children, EliminateProjection(child.(PhysicalPlan)))
+// canLogicalProjectionBeEliminated checks if a PROJECTION in a logical plan can be eliminated.
+// Projection p can be eliminated when all the following conditions are satisfied:
+// 1. it's not the root operator
+// 2. canProjectionBeEliminated(p) returns true
+func canLogicalProjectionBeEliminated(p *Projection) bool {
+	if len(p.Parents()) == 0 {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: If the query is `select * from t limit 1`, the parents will also be 0 and can't be removed.

File: plan/eliminate_projection.go [hunk 5]
```diff
@@ -14,50 +14,215 @@
...
+		projCol, ok := expr.(*expression.Column)
+		if !ok || !projCol.Equal(childCols[i], nil) {
 			return false
 		}
 	}
 	return true
 }
+
+func resolveColumnAndReplace(origin *expression.Column, replace map[string]*expression.Column) {
+	dst := replace[string((*origin).HashCode())]
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Needn't be (*orgin)

File: plan/eliminate_projection.go (hunk 4 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: yes, so it returns false

File: plan/eliminate_projection.go [hunk 6]
```diff
@@ -14,50 +14,215 @@
...
+		for i, parentColumn := range plan.Schema().Columns {
+			childColumns[i].ColName = parentColumn.ColName
+		}
+		RemovePlan(p)
+		return child
+	}
+	return p
+}
+
+type nonRootProjectionEliminater struct {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Needn't call it `nonRoot...`, it's confusing. ProjectonElimiater is ok.

File: plan/eliminate_projection.go [hunk 7]
```diff
@@ -14,50 +14,215 @@
...
+	p.LogicalJoin.replaceExprColumns(replace)
+	for _, coCol := range p.corCols {
+		dst := replace[string(coCol.Column.HashCode())]
+		if dst != nil {
+			coCol.Column = *dst
+		}
+	}
+}
+
+func (p *DataSource) replaceExprColumns(replace map[string]*expression.Column) {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: I think DataSource needn't implement it.

File: plan/eliminate_projection.go [hunk 8]
```diff
@@ -14,50 +14,215 @@
...
+	}
+}
+
+func (p *TopN) replaceExprColumns(replace map[string]*expression.Column) {
+	for _, byItem := range p.ByItems {
+		resolveExprAndReplace(byItem.Expr, replace)
+	}
+}
+
+func (p *Update) replaceExprColumns(replace map[string]*expression.Column) {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Update also needn't

File: plan/eliminate_projection.go [hunk 9]
```diff
@@ -14,50 +14,215 @@
...
+			resolveExprAndReplace(aggExpr, replace)
+		}
+	}
+	for _, gbyItem := range p.GroupByItems {
+		resolveExprAndReplace(gbyItem, replace)
+	}
+	for _, gbyCol := range p.groupByCols {
+		resolveColumnAndReplace(gbyCol, replace)
+	}
+	for _, prop := range p.possibleProperties {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: needn't replace it

File: plan/eliminate_projection.go [hunk 10]
```diff
@@ -14,50 +14,215 @@
...
+func (p *LogicalAggregation) replaceExprColumns(replace map[string]*expression.Column) {
+	for _, agg := range p.AggFuncs {
+		for _, aggExpr := range agg.GetArgs() {
+			resolveExprAndReplace(aggExpr, replace)
+		}
+	}
+	for _, gbyItem := range p.GroupByItems {
+		resolveExprAndReplace(gbyItem, replace)
+	}
+	for _, gbyCol := range p.groupByCols {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Just recollect it, needn't replace it

File: plan/eliminate_projection.go [hunk 11]
```diff
@@ -14,50 +14,215 @@
...
+		for _, prop := range leftProp {
+			resolveColumnAndReplace(prop, replace)
+		}
+	}
+	for _, rightProp := range p.rightProperties {
+		for _, prop := range rightProp {
+			resolveColumnAndReplace(prop, replace)
+		}
+	}
+	if p.redundantSchema != nil {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Needn't replace it, it's only needed by logicalplanbuilder

File: plan/eliminate_projection.go (hunk 6 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: After refactoring, projection elimination is done in both logical and physical optimization, but they are different.

In logical optimization, we only eliminate the no-root projections, leave the root projection, if exists and can be eliminated, to be eliminated in the physical optimization

I hope to use these abnormal function and struct names, which are `nonRootProjectionEliminater` and `eliminateRootProjection`, to inform people the difference between them.

File: plan/eliminate_projection.go (hunk 5 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: fixed

File: plan/eliminate_projection.go (hunk 7 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: removed

File: plan/eliminate_projection.go (hunk 8 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: removed

File: plan/eliminate_projection.go (hunk 9 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: removed

File: plan/eliminate_projection.go (hunk 10 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: done

File: plan/eliminate_projection.go (hunk 11 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: removed

File: plan/eliminate_projection.go (hunk 4 above)
Comment (hanfei1991) [inline plan/eliminate_projection.go]: the parentes length will be 1 but can't be removed....

File: plan/eliminate_projection.go [hunk 12]
```diff
@@ -14,50 +14,185 @@
...
+	for _, otherExpr := range p.OtherConditions {
+		resolveExprAndReplace(otherExpr, replace)
+	}
+	for _, leftKey := range p.LeftJoinKeys {
+		resolveColumnAndReplace(leftKey, replace)
+	}
+	for _, rightKey := range p.RightJoinKeys {
+		resolveColumnAndReplace(rightKey, replace)
+	}
+	for _, leftProp := range p.leftProperties {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: No need 2 update properties

File: plan/physical_plan_builder.go [hunk 13]
```diff
@@ -130,40 +131,41 @@ func (p *DataSource) convert2IndexScan(prop *requiredProperty, index *model.Inde
...
-			is.IndexConditionPBExpr, is.indexFilterConditions, idxConds = expression.ExpressionsToPB(sc, idxConds, client)
-			is.TableConditionPBExpr, is.tableFilterConditions, tblConds = expression.ExpressionsToPB(sc, tblConds, client)
-			newSel.Conditions = append(idxConds, tblConds...)
-		}
-		var err error
-		is.Ranges, err = ranger.BuildIndexRange(p.ctx.GetSessionVars().StmtCtx, is.Table, is.Index, is.accessInAndEqCount, is.AccessCondition)
-		if err != nil {
-			if !terror.ErrorEqual(err, types.ErrTruncated) {
+	is.Ranges = ranger.FullIndexRange()
+	if len(p.parents) > 0 {
```
Comment (hanfei1991) [inline plan/physical_plan_builder.go]: if len(p.parents) > 0 && ...

File: plan/eliminate_projection.go [hunk 14]
```diff
@@ -14,50 +14,245 @@
...
+// eliminate eliminates the redundant projection in a logical plan.
+// The order of output plan's schema columns can be changed if "haveOrderKeeper" is true.
+func (pe *projectionEliminater) eliminate(p Plan, replace map[string]*expression.Column, haveOrderKeeper bool) Plan {
+	proj, isProj := p.(*Projection)
+	_, isJoin := p.(*LogicalJoin)
+	_, isAggr := p.(*LogicalAggregation)
+	isOrderKeeper := isProj || isJoin || isAggr
+
+	_, isUnion := p.(*Union)
+	_, isSort := p.(*Sort)
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: It seems we need't to check sort

File: plan/eliminate_projection.go [hunk 15]
```diff
@@ -14,50 +14,245 @@
...
+func (pe *projectionEliminater) optimize(lp LogicalPlan, _ context.Context, _ *idAllocator) (LogicalPlan, error) {
+	root := pe.eliminate(lp, make(map[string]*expression.Column), false)
+	return root.(LogicalPlan), nil
+}
+
+// eliminate eliminates the redundant projection in a logical plan.
+// The order of output plan's schema columns can be changed if "haveOrderKeeper" is true.
+func (pe *projectionEliminater) eliminate(p Plan, replace map[string]*expression.Column, haveOrderKeeper bool) Plan {
+	proj, isProj := p.(*Projection)
+	_, isJoin := p.(*LogicalJoin)
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: It seems we needn't to check join and projection.

File: plan/eliminate_projection.go [hunk 16]
```diff
@@ -14,50 +14,245 @@
...
 }
+
+func resolveColumnAndReplace(origin *expression.Column, replace map[string]*expression.Column) {
+	dst := replace[string(origin.HashCode())]
+	if dst != nil {
+		*origin = *dst
+	}
+}
+
+func resolveExprAndReplace(origin expression.Expression, replace map[string]*expression.Column) {
```
Comment (zz-jason) [inline plan/eliminate_projection.go]: As previous discussion, function `expression.ColumnSubstitute(expr Expression, schema *Schema, newExprs []Expression)` is too heavy and inefficient: 
 1. for every `ScalarFunction` node in the expression tree, this function will do a copy for it.
 2. uses a loop to find whether a column exists in `schema *Schema`

I prefer not to use it

File: plan/eliminate_projection.go [hunk 17]
```diff
@@ -14,50 +14,257 @@
...
+	case *expression.ScalarFunction:
+		for _, arg := range expr.GetArgs() {
+			resolveExprAndReplace(arg, replace)
+		}
+	}
+}
+
+// eliminatePhysicalProjection should be called after physical optimization to eliminate the redundant projection
+// left after logical projection elimination.
+func eliminatePhysicalProjection(p Plan, replace map[string]*expression.Column) Plan {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Restrict p as PhysicalPlan.

File: plan/eliminate_projection.go [hunk 18]
```diff
@@ -14,50 +14,257 @@
...
+
+// optimize implements the logicalOptRule interface
+func (pe *projectionEliminater) optimize(lp LogicalPlan, _ context.Context, _ *idAllocator) (LogicalPlan, error) {
+	root := pe.eliminate(lp, make(map[string]*expression.Column), false)
+	return root.(LogicalPlan), nil
+}
+
+// eliminate eliminates the redundant projection in a logical plan.
+// The order of output plan's schema columns can be changed if "haveProjection" is true.
+func (pe *projectionEliminater) eliminate(p Plan, replace map[string]*expression.Column, haveProjection bool) Plan {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Restrict p as Logical Plan

File: plan/eliminate_projection.go [hunk 19]
```diff
@@ -14,50 +14,257 @@
...
+
+	_, isSort := p.(*Sort)
+	_, isTopN := p.(*TopN)
+	_, isLimit := p.(*Limit)
+	_, isSelect := p.(*Selection)
+	_, isMaxOneRow := p.(*MaxOneRow)
+
+	_, isApply := p.(*LogicalApply)
+	_, isJoin := p.(*LogicalJoin)
+	if isSort || isTopN || isLimit || isSelect || isMaxOneRow {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: switch x := p.(type) {case *Sort, *TopN ...:} Make sure you don't forget any type by checking this file : https://github.com/pingcap/tidb/blob/master/plan/logical_plans.go

File: plan/eliminate_projection.go [hunk 20]
```diff
@@ -14,50 +14,257 @@
...
+		p.SetSchema(p.Children()[0].Schema())
+	} else if isApply || isJoin {
+		var joinTp JoinType
+		if isApply {
+			joinTp = p.(*LogicalApply).JoinType
+		} else {
+			joinTp = p.(*LogicalJoin).JoinType
+		}
+		switch joinTp {
+		case InnerJoin:
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: case InnerJoin, LeftOuterJoin, RightOuterJoin blabla...

File: plan/eliminate_projection.go [hunk 21]
```diff
@@ -14,50 +14,257 @@
...
+		case LeftOuterSemiJoin:
+			newSchema := p.Children()[0].Schema().Clone()
+			newSchema.Append(p.Schema().Columns[len(p.Schema().Columns)-1])
+			p.SetSchema(newSchema)
+		}
+	} else {
+		for _, dst := range p.Schema().Columns {
+			resolveColumnAndReplace(dst, replace)
+		}
+		for _, key := range p.Schema().Keys {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: keys haven't been builded.

File: plan/eliminate_projection.go [hunk 22]
```diff
@@ -14,50 +14,257 @@
...
+			}
+		}
+	}
+	p.replaceExprColumns(replace)
+
+	if !isProj {
+		return p
+	}
+
+	child := p.Children()[0]
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Why check child is join ?

Comment (zz-jason) [inline plan/eliminate_projection.go]: to guarantee there is a projection between root/union and join

File: plan/eliminate_projection.go (hunk 17 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: done

File: plan/eliminate_projection.go (hunk 18 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: done

File: plan/eliminate_projection.go (hunk 19 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: done

File: plan/eliminate_projection.go (hunk 20 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: done

File: plan/eliminate_projection.go (hunk 21 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: removed

File: plan/eliminate_projection.go [hunk 23]
```diff
@@ -14,50 +14,229 @@
...
+}
+
+// eliminate eliminates the redundant projection in a logical plan.
+// The order of output plan's schema columns can be changed if "haveProjection" is true.
+func (pe *projectionEliminater) eliminate(p LogicalPlan, replace map[string]*expression.Column, haveProjection bool) LogicalPlan {
+	proj, isProj := p.(*Projection)
+	children := make([]Plan, 0, len(p.Children()))
+	_, isUnion := p.(*Union)
+	for _, child := range p.Children() {
+		children = append(children, pe.eliminate(child.(LogicalPlan), replace, !isUnion && (haveProjection || isProj)))
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Well, I think there are two points that can be improved:
1. haveProjection is not a good name.
2. !isUnion && (haveProjection || isProj) is not readable. 
The following style is more acceptable:
```
var childFlag = haveProjection
if isProj {
    childFlag = true
} else if _, isUnion := p.(*Union); isUnion {
    childFlag = false
}
```

File: plan/eliminate_projection.go [hunk 24]
```diff
@@ -14,50 +14,229 @@
...
+	if !(isProj && haveProjection && canProjectionBeEliminatedLoose(proj)) {
+		return p
+	}
+
+	child := p.Children()[0]
+	colNameMap := make(map[string]model.CIStr)
+	exprs := proj.Exprs
+	for i, parentColumn := range proj.Schema().Columns {
+		col, _ := exprs[i].(*expression.Column)
+		col.DBName = parentColumn.DBName
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Why do this replacement ? It doesn't do favor for execution.

File: plan/eliminate_projection.go [hunk 25]
```diff
@@ -14,50 +14,229 @@
...
+		col.TblName = parentColumn.TblName
+		col.ColName = parentColumn.ColName
+		replace[string(parentColumn.HashCode())] = col
+		colNameMap[string(col.HashCode())] = parentColumn.ColName
+	}
+	for _, childCol := range child.Schema().Columns {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: ditto.

File: plan/eliminate_projection.go [hunk 26]
```diff
@@ -14,50 +14,229 @@
...
+	p.replaceExprColumns(replace)
+
+	proj, isProj := p.(*Projection)
+	if !isProj || !canProjectionBeEliminatedStrict(proj) {
+		return p
+	}
+	child := p.Children()[0]
+	exprs := proj.Exprs
+	childCols := child.Schema().Columns
+	for i, parentColumn := range proj.Schema().Columns {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: I think only root plan needs to refresh his child's schema.

File: plan/eliminate_projection.go [hunk 27]
```diff
@@ -14,50 +14,229 @@
...
+	children := make([]Plan, 0, len(p.Children()))
+	for _, child := range p.Children() {
+		children = append(children, eliminatePhysicalProjection(child.(PhysicalPlan), replace))
+	}
+	p.SetChildren(children...)
+
+	for _, dst := range p.Schema().Columns {
+		resolveColumnAndReplace(dst, replace)
+	}
+	for _, key := range p.Schema().Keys {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: Key Info is useless afterwards.

File: plan/eliminate_projection.go [hunk 28]
```diff
@@ -14,50 +14,229 @@
...
+			p.SetSchema(expression.MergeSchema(p.Children()[0].Schema(), p.Children()[1].Schema()))
+		case SemiJoin:
+			p.SetSchema(p.Children()[0].Schema().Clone())
+		case LeftOuterSemiJoin:
+			newSchema := p.Children()[0].Schema().Clone()
+			newSchema.Append(p.Schema().Columns[len(p.Schema().Columns)-1])
+			p.SetSchema(newSchema)
+		}
+	default:
+		for _, dst := range p.Schema().Columns {
```
Comment (zz-jason) [inline plan/eliminate_projection.go]: `LogicalAggregation` may relying on its child

File: plan/eliminate_projection.go (hunk 26 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: No. actually, we only need to refresh child's schema in physical projection elimination when the eliminated projection is root.

File: plan/eliminate_projection.go (hunk 28 above)
Comment (hanfei1991) [inline plan/eliminate_projection.go]: LogicalAggregation's schema is decide by its own aggregate functions.

File: plan/eliminate_projection.go [hunk 29]
```diff
@@ -14,50 +14,230 @@
...
+}
+
+// optimize implements the logicalOptRule interface
+func (pe *projectionEliminater) optimize(lp LogicalPlan, _ context.Context, _ *idAllocator) (LogicalPlan, error) {
+	root := pe.eliminate(lp, make(map[string]*expression.Column), false)
+	return root.(LogicalPlan), nil
+}
+
+// eliminate eliminates the redundant projection in a logical plan.
+func (pe *projectionEliminater) eliminate(p LogicalPlan, replace map[string]*expression.Column, canEliminate bool) LogicalPlan {
```
Comment (hanfei1991) [inline plan/eliminate_projection.go]: It is better to make replace as a return value.

File: plan/eliminate_projection.go (hunk 27 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: removed

**Comment by hanfei1991** [comment]
LGTM

**Comment by hanfei1991** [comment]
@winoros @lamxTyler PTAL

File: plan/eliminate_projection.go (hunk 28 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: Aggregation function "firstrow"'s corresponding output column is cloned from its child's. the `default` branch has to be retained.

File: plan/optimizer.go [hunk 30]
```diff
@@ -119,10 +121,24 @@ func doOptimize(flag uint64, logic LogicalPlan, ctx context.Context, allocator *
...
+	if oldRoot.ID() != finalPlan.ID() {
+		newCols := finalPlan.Schema().Columns
+		for i, oldCol := range oldRoot.Schema().Columns {
+			newCols[i].ColName = oldCol.ColName
+		}
 	}
-	return physicalOptimize(flag, logic, allocator)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
```
Comment (lamxTyler) [inline plan/optimizer.go]: Can it be put before line 130?

File: plan/eliminate_projection.go [hunk 31]
```diff
@@ -14,50 +14,225 @@
...
+			newCols[i].ColName = oldCol.ColName
+		}
+	}
+	return newRoot
+}
+
+type projectionEliminater struct {
+}
+
+// optimize implements the logicalOptRule interface
```
Comment (lamxTyler) [inline plan/eliminate_projection.go]: Add `.` at the end.

File: plan/plan.go [hunk 32]
```diff
@@ -61,6 +61,8 @@ type Plan interface {
 	SetParents(...Plan)
 	// SetChildren sets the children for the plan.
 	SetChildren(...Plan)
+	// replaceEpxrColumns replace all the column reference in the plan's expression node
```
Comment (lamxTyler) [inline plan/plan.go]: Add `.`.

Comment (lamxTyler) [inline plan/plan.go]: `replaceEpxr...` -> replaceExpr...`

Comment (zz-jason) [inline plan/plan.go]: done

File: plan/eliminate_projection.go (hunk 31 above)
Comment (zz-jason) [inline plan/eliminate_projection.go]: done

File: plan/optimizer.go (hunk 30 above)
Comment (zz-jason) [inline plan/optimizer.go]: moved

**Pull request merged**